From ea32d2faaa11644d3a4dba3cf284df4c29ec7f1b Mon Sep 17 00:00:00 2001
From: Evoke Zhang <evoke.zhang@amlogic.com>
Date: Mon, 6 Jan 2014 14:40:47 +0800
Subject: [PATCH 2706/5965] PD #85473: add lcd external config (default
 disable) in basic board dt for reference, disable debug print

---
 arch/arm/boot/dts/amlogic/aml_top.dtd         |   2 +
 arch/arm/boot/dts/amlogic/meson8_k100_1G.dtd  |  17 ++
 arch/arm/boot/dts/amlogic/meson8_k100_v1.dtd  |  17 ++
 arch/arm/boot/dts/amlogic/meson8_k101_v1.dtd  |  19 ++
 arch/arm/boot/dts/amlogic/meson8_k102_v1.dtd  |  19 ++
 drivers/amlogic/display/vout/Kconfig          |   7 +
 drivers/amlogic/display/vout/Makefile         |   1 +
 drivers/amlogic/display/vout/aml_lcd.dtd      |  54 ++++-
 .../amlogic/display/vout/lcd_extern/Kconfig   |  12 +
 .../amlogic/display/vout/lcd_extern/Makefile  |   4 +
 .../amlogic/display/vout/lcd_extern/aml.dtd   |  21 ++
 .../display/vout/lcd_extern/lcd_extern.c      | 219 ++++++++++++++++++
 .../amlogic/display/vout/lcd_extern/tc101.c   | 211 +++++++++++++++++
 drivers/amlogic/display/vout/lcdoutc.c        |  33 ++-
 include/linux/amlogic/vout/aml_lcd_extern.h   |  35 +++
 15 files changed, 666 insertions(+), 5 deletions(-)
 create mode 100755 drivers/amlogic/display/vout/lcd_extern/Kconfig
 create mode 100755 drivers/amlogic/display/vout/lcd_extern/Makefile
 create mode 100755 drivers/amlogic/display/vout/lcd_extern/aml.dtd
 create mode 100755 drivers/amlogic/display/vout/lcd_extern/lcd_extern.c
 create mode 100755 drivers/amlogic/display/vout/lcd_extern/tc101.c
 create mode 100755 include/linux/amlogic/vout/aml_lcd_extern.h

diff --git a/arch/arm/boot/dts/amlogic/aml_top.dtd b/arch/arm/boot/dts/amlogic/aml_top.dtd
index e18cdc5902b3..2c635f911131 100755
--- a/arch/arm/boot/dts/amlogic/aml_top.dtd
+++ b/arch/arm/boot/dts/amlogic/aml_top.dtd
@@ -258,6 +258,8 @@ sub_file /drivers/amlogic/usb/aml.dtd
 sub_file /drivers/amlogic/display/backlight/aml.dtd
 #device lcd
 sub_file /drivers/amlogic/display/vout/aml_lcd.dtd
+#device lcd_extern
+sub_file /drivers/amlogic/display/vout/lcd_extern/aml.dtd
 
 /// ***************************************************************************************
 ///     -       Sensors
diff --git a/arch/arm/boot/dts/amlogic/meson8_k100_1G.dtd b/arch/arm/boot/dts/amlogic/meson8_k100_1G.dtd
index 34970573fdc1..4f17efeee017 100755
--- a/arch/arm/boot/dts/amlogic/meson8_k100_1G.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_k100_1G.dtd
@@ -1129,7 +1129,24 @@ sdio{
 		pol_hsync_vsync=<0 0>;	/** hsync_polarity, vsync_polarity : (0=negative, 1=positive) */
 	};
 
+//$$ DEVICE="lcd_extern"
+//$$ L2 PROP_STR = "status"
+	lcd_extern{
+		compatible = "amlogic,lcd_extern";
+		status = "disabled";
 
+//$$ DEVICE = "lcd_extern_tc101"
+//$$ L2 PROP_STR = "i2c_bus"
+//$$ L2 PROP_U32 = "reg"
+//$$ L2 PROP_STR = "status"
+		lcd_extern_tc101{
+			dev_name ="lcd_tc101";
+			type = <0>;	/** lcd_extern_driver type: 0=i2c, 1=spi */
+			address = <0x7e>;
+			i2c_bus = "i2c_bus_a";
+			status = "disabled";
+		};
+	};
 
 /// ***************************************************************************************
 ///     -       Sensors
diff --git a/arch/arm/boot/dts/amlogic/meson8_k100_v1.dtd b/arch/arm/boot/dts/amlogic/meson8_k100_v1.dtd
index ccf74be30543..adca52a4ba49 100755
--- a/arch/arm/boot/dts/amlogic/meson8_k100_v1.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_k100_v1.dtd
@@ -1141,7 +1141,24 @@ sdio{
 		pol_hsync_vsync=<0 0>;	/** hsync_polarity, vsync_polarity : (0=negative, 1=positive) */
 	};
 
+//$$ DEVICE="lcd_extern"
+//$$ L2 PROP_STR = "status"
+	lcd_extern{
+		compatible = "amlogic,lcd_extern";
+		status = "disabled";
 
+//$$ DEVICE = "lcd_extern_tc101"
+//$$ L2 PROP_STR = "i2c_bus"
+//$$ L2 PROP_U32 = "reg"
+//$$ L2 PROP_STR = "status"
+		lcd_extern_tc101{
+			dev_name ="lcd_tc101";
+			type = <0>;	/** lcd_extern_driver type: 0=i2c, 1=spi */
+			address = <0x7e>;
+			i2c_bus = "i2c_bus_a";
+			status = "disabled";
+		};
+	};
 
 /// ***************************************************************************************
 ///     -       Sensors
diff --git a/arch/arm/boot/dts/amlogic/meson8_k101_v1.dtd b/arch/arm/boot/dts/amlogic/meson8_k101_v1.dtd
index 36fef1dda904..0feebc12adaf 100755
--- a/arch/arm/boot/dts/amlogic/meson8_k101_v1.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_k101_v1.dtd
@@ -1138,6 +1138,25 @@ sdio{
     factor = < 2 3 >;
     lane_num=<4>;//lane num
 	};
+
+//$$ DEVICE="lcd_extern"
+//$$ L2 PROP_STR = "status"
+	lcd_extern{
+		compatible = "amlogic,lcd_extern";
+		status = "disabled";
+
+//$$ DEVICE = "lcd_extern_tc101"
+//$$ L2 PROP_STR = "i2c_bus"
+//$$ L2 PROP_U32 = "reg"
+//$$ L2 PROP_STR = "status"
+		lcd_extern_tc101{
+			dev_name ="lcd_tc101";
+			type = <0>;	/** lcd_extern_driver type: 0=i2c, 1=spi */
+			address = <0x7e>;
+			i2c_bus = "i2c_bus_a";
+			status = "disabled";
+		};
+	};
 /// ***************************************************************************************
 ///     -       Sensors
 //$$ MODULE="Sensors"
diff --git a/arch/arm/boot/dts/amlogic/meson8_k102_v1.dtd b/arch/arm/boot/dts/amlogic/meson8_k102_v1.dtd
index c229ae6fe85c..eb0f9038d7c2 100755
--- a/arch/arm/boot/dts/amlogic/meson8_k102_v1.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_k102_v1.dtd
@@ -1157,6 +1157,25 @@ sdio{
     factor = < 2 3 >;
     lane_num=<4>;//lane num
 	};
+
+//$$ DEVICE="lcd_extern"
+//$$ L2 PROP_STR = "status"
+	lcd_extern{
+		compatible = "amlogic,lcd_extern";
+		status = "disabled";
+
+//$$ DEVICE = "lcd_extern_tc101"
+//$$ L2 PROP_STR = "i2c_bus"
+//$$ L2 PROP_U32 = "reg"
+//$$ L2 PROP_STR = "status"
+		lcd_extern_tc101{
+			dev_name ="lcd_tc101";
+			type = <0>;	/** lcd_extern_driver type: 0=i2c, 1=spi */
+			address = <0x7e>;
+			i2c_bus = "i2c_bus_a";
+			status = "disabled";
+		};
+	};
 /// ***************************************************************************************
 ///     -       Sensors
 //$$ MODULE="Sensors"
diff --git a/drivers/amlogic/display/vout/Kconfig b/drivers/amlogic/display/vout/Kconfig
index d150c4c682db..bbe5b0646512 100755
--- a/drivers/amlogic/display/vout/Kconfig
+++ b/drivers/amlogic/display/vout/Kconfig
@@ -15,6 +15,13 @@ config AM_LCD_OUTPUT
 	help  
 		LCD output module
 
+config AML_LCD_EXTERN
+	tristate "AML lcd external bridge driver"
+	default y
+	depends on  AM_LCD_OUTPUT
+	help
+		AML lcd external bridge driver		
+		
 config AML_GAMMA_DEBUG
 	tristate "AML Gamma debug"
 	default y
diff --git a/drivers/amlogic/display/vout/Makefile b/drivers/amlogic/display/vout/Makefile
index 3df2d2ec90ff..68bf8c2f9e23 100755
--- a/drivers/amlogic/display/vout/Makefile
+++ b/drivers/amlogic/display/vout/Makefile
@@ -9,5 +9,6 @@ am_tvout-objs :=   tvoutc.o  tvconf.o enc_clk_config.o
 obj-$(CONFIG_AM_LCD_OUTPUT)  += lcdoutc.o
 obj-$(CONFIG_AM_LCD_OUTPUT)  += edp_drv.o
 obj-$(CONFIG_AM_LCD_OUTPUT)  += mipi_dsi_util.o
+obj-$(CONFIG_AM_LCD_OUTPUT)	+= lcd_extern/
 
 #EXTRA_CFLAGS += -DCONFIG_PM
diff --git a/drivers/amlogic/display/vout/aml_lcd.dtd b/drivers/amlogic/display/vout/aml_lcd.dtd
index 3c03a7fa740e..33a86c5960c4 100755
--- a/drivers/amlogic/display/vout/aml_lcd.dtd
+++ b/drivers/amlogic/display/vout/aml_lcd.dtd
@@ -330,7 +330,7 @@
 //$$ L2 PROP_U32 2 = "hsync_width_backporch"
 //$$ L2 PROP_U32 2 = "vsync_width_backporch"
 //$$ L2 PROP_U32 2 = "pol_hsync_vsync"
-lcd_HE050NA01F:lcd_HE050NA01F{	//800*480
+lcd_HE050NA01F:lcd_HE050NA01F{
 		model_name="HE050NA01F";	/** lcd model name */
 		interface="ttl";		/** lcd_interface(mipi, lvds, edp, ttl) */
 		active_area=<108 65>;	/** lcd active_area or display_area horizontal size, vertical size : (unit in mm, you can find them on the home page of lcd spec) */
@@ -344,5 +344,57 @@ lcd_HE050NA01F:lcd_HE050NA01F{	//800*480
 		vsync_width_backporch=<5 23>;	/** vsync_width, vsync_backporch(include vsync_width) */
 		pol_hsync_vsync=<0 0>;	/** hsync_polarity, vsync_polarity : (0=negative, 1=positive) */
 	};
+	
+//$$ MATCH "lcd_model_config_match" = <&lcd_HSD043I9W1>
+//$$ L2 PROP_STR = "model_name"
+//$$ L2 PROP_STR = "interface"
+//$$ L2 PROP_U32 2 = "active_area"
+//$$ L2 PROP_U32 2 = "lcd_bits_option"
+//$$ L2 PROP_U32 2 = "resolution"
+//$$ L2 PROP_U32 2 = "period"
+//$$ L2 PROP_U32 2 = "clock_hz_pol"
+//$$ L2 PROP_U32 2 = "hsync_width_backporch"
+//$$ L2 PROP_U32 2 = "vsync_width_backporch"
+//$$ L2 PROP_U32 2 = "pol_hsync_vsync"		
+	lcd_HSD043I9W1:lcd_HSD043I9W1{
+		model_name="HSD043I9W1";	/** lcd model name */
+		interface="ttl";		/** lcd_interface(mipi, lvds, edp, ttl) */
+		active_area=<95 54>;	/** lcd active_area or display_area horizontal size, vertical size : (unit in mm, you can find them on the home page of lcd spec) */
+		lcd_bits_option=<8 1>;	/** lcd_bits(6,8), bits_option(0=only support one mode as lcd_bits define, 1=both support 6/8bit) */
+		
+		resolution=<480 272>;	/** horizontal resolution, vertical resolution */
+		period=<620 310>;		/** horizontal period(htotal), vertical period(vtotal) */
+		
+		clock_hz_pol=<11500000 0>;	/** clock_hz(unit in Hz, both support clk and frame_rate, >200 regard as clk, <200 regard as frame_rate), clk_pol(only valid for TTL) */
+		hsync_width_backporch=<5 40>;	/** hsync_width, hsync_backporch(include hsync_width) */
+		vsync_width_backporch=<2 8>;	/** vsync_width, vsync_backporch(include vsync_width) */
+		pol_hsync_vsync=<0 0>;	/** hsync_polarity, vsync_polarity : (0=negative, 1=positive) */
+	};
+	
+//$$ MATCH "lcd_model_config_match" = <&lcd_TM080XSZ00>
+//$$ L2 PROP_STR = "model_name"
+//$$ L2 PROP_STR = "interface"
+//$$ L2 PROP_U32 2 = "active_area"
+//$$ L2 PROP_U32 2 = "lcd_bits_option"
+//$$ L2 PROP_U32 2 = "resolution"
+//$$ L2 PROP_U32 2 = "period"
+//$$ L2 PROP_U32 2 = "clock_hz_pol"
+//$$ L2 PROP_U32 2 = "hsync_width_backporch"
+//$$ L2 PROP_U32 2 = "vsync_width_backporch"
+//$$ L2 PROP_U32 2 = "pol_hsync_vsync"
+lcd_TM080XSZ00:lcd_TM080XSZ00{
+		model_name="TM080XSZ00";	/** lcd model name */
+		interface="ttl";		/** lcd_interface(mipi, lvds, edp, ttl) */
+		active_area=<174 104>;	/** lcd active_area or display_area horizontal size, vertical size : (unit in mm, you can find them on the home page of lcd spec) */
+		lcd_bits_option=<6 0>;	/** lcd_bits(6,8), bits_option(0=only support one mode as lcd_bits define, 1=both support 6/8bit) */
+		
+		resolution=<1280 768>;	/** horizontal resolution, vertical resolution */
+		period=<1440 810>;		/** horizontal period(htotal), vertical period(vtotal) */
+		
+		clock_hz_pol=<70000000 0>;	/** clock_hz(unit in Hz, both support clk and frame_rate, >200 regard as clk, <200 regard as frame_rate), clk_pol(only valid for TTL) */
+		hsync_width_backporch=<10 70>;	/** hsync_width, hsync_backporch(include hsync_width) */
+		vsync_width_backporch=<2 20>;	/** vsync_width, vsync_backporch(include vsync_width) */
+		pol_hsync_vsync=<0 0>;	/** hsync_polarity, vsync_polarity : (0=negative, 1=positive) */
+	};
 
 #endif
diff --git a/drivers/amlogic/display/vout/lcd_extern/Kconfig b/drivers/amlogic/display/vout/lcd_extern/Kconfig
new file mode 100755
index 000000000000..7aab495c7283
--- /dev/null
+++ b/drivers/amlogic/display/vout/lcd_extern/Kconfig
@@ -0,0 +1,12 @@
+#
+# Video output configuration
+#
+menu "Amlogic LCD External driver"
+
+config AML_LCD_EXTERN
+	tristate "LCD External driver"
+	default y
+	help  
+		lcd exteranl driver
+
+endmenu
diff --git a/drivers/amlogic/display/vout/lcd_extern/Makefile b/drivers/amlogic/display/vout/lcd_extern/Makefile
new file mode 100755
index 000000000000..c4d1f48d969e
--- /dev/null
+++ b/drivers/amlogic/display/vout/lcd_extern/Makefile
@@ -0,0 +1,4 @@
+
+obj-$(CONFIG_AML_LCD_EXTERN)	+= lcd_extern.o
+obj-$(CONFIG_AML_LCD_EXTERN)	+= tc101.o
+
diff --git a/drivers/amlogic/display/vout/lcd_extern/aml.dtd b/drivers/amlogic/display/vout/lcd_extern/aml.dtd
new file mode 100755
index 000000000000..5f207a1a0b23
--- /dev/null
+++ b/drivers/amlogic/display/vout/lcd_extern/aml.dtd
@@ -0,0 +1,21 @@
+#ifdef CONFIG_AML_LCD_EXTERN
+//$$ DEVICE="lcd_extern"
+//$$ L2 PROP_STR = "status"
+	lcd_extern{
+		compatible = "amlogic,lcd_extern";
+		status = "disabled";
+
+//$$ DEVICE = "lcd_extern_tc101"
+//$$ L2 PROP_STR = "i2c_bus"
+//$$ L2 PROP_U32 = "reg"
+//$$ L2 PROP_STR = "status"
+		lcd_extern_tc101{
+			dev_name ="lcd_tc101";
+			type = <0>;	/** lcd_extern_driver type: 0=i2c, 1=spi */
+			address = <0x7e>;
+			i2c_bus = "i2c_bus_a";
+			status = "disabled";
+		};
+	};
+
+#endif
diff --git a/drivers/amlogic/display/vout/lcd_extern/lcd_extern.c b/drivers/amlogic/display/vout/lcd_extern/lcd_extern.c
new file mode 100755
index 000000000000..b30060320f8e
--- /dev/null
+++ b/drivers/amlogic/display/vout/lcd_extern/lcd_extern.c
@@ -0,0 +1,219 @@
+#include <linux/kernel.h>
+#include <linux/module.h>
+#include <linux/interrupt.h>
+#include <linux/platform_device.h>
+#include <linux/i2c.h>
+#include <linux/i2c-aml.h>
+#include <linux/clk.h>
+#include <linux/delay.h>
+#include <linux/slab.h>
+#include <linux/of.h>
+#include <mach/am_regs.h>
+#include <mach/gpio.h>
+#include <linux/amlogic/vout/aml_lcd_extern.h>
+#include <linux/amlogic/aml_gpio_consumer.h>
+
+//#define LCD_EXT_DEBUG_INFO
+#ifdef LCD_EXT_DEBUG_INFO
+#define DBG_PRINT(...)		printk(__VA_ARGS__)
+#else
+#define DBG_PRINT(...)
+#endif
+
+struct aml_lcd_extern_driver_t lcd_ext_driver;
+
+struct aml_lcd_extern_driver_t* aml_lcd_extern_get_driver(void)
+{
+	return &lcd_ext_driver;
+}
+
+static int get_lcd_extern_dt_data(struct device_node* of_node, struct lcd_extern_data_t *pdata)
+{
+	int err;
+	const char *str;
+	
+	err = of_property_read_string(of_node, "dev_name", (const char **)&pdata->name);
+	if (err) {
+		pdata->name = "aml_lcd_extern";
+		printk("warning: get dev_name failed, set dev_name aml_lcd_extern\n");
+	}
+	err = of_property_read_u32(of_node, "type", &pdata->type);
+	if (err) {
+		pdata->type = LCD_EXTERN_MAX;
+		printk("warning: get type failed, exit\n");
+		return -1;
+	}
+	err = of_property_read_string(of_node, "status", &str);
+	if (err) {
+		printk("%s warning: get status failed, default disable\n", pdata->name);
+		pdata->status = 0;
+	}
+	else {
+		if ((strncmp(str, "ok", 2) == 0) || (strncmp(str, "enable", 6) == 0))
+			pdata->status = 1;
+		else
+			pdata->status = 0;
+	}
+	switch (pdata->type) {
+		case LCD_EXTERN_I2C:
+			err = of_property_read_u32(of_node,"address",&pdata->addr);
+			if (err) {
+				printk("%s warning: get i2c address failed\n", pdata->name);
+				pdata->addr = 0;
+			}
+			DBG_PRINT("%s: address=0x%02x\n", pdata->name, pdata->addr);
+		  
+			err = of_property_read_string(of_node, "i2c_bus", &str);
+			if (err) {
+				printk("%s warning: get i2c_bus failed, use default i2c bus\n", pdata->name);
+				pdata->i2c_bus = AML_I2C_MASTER_A;
+			}
+			else {
+				if (strncmp(str, "i2c_bus_a", 9) == 0)
+					pdata->i2c_bus = AML_I2C_MASTER_A;
+				else if (strncmp(str, "i2c_bus_b", 9) == 0)
+					pdata->i2c_bus = AML_I2C_MASTER_B;
+				else if (strncmp(str, "i2c_bus_c", 9) == 0)
+					pdata->i2c_bus = AML_I2C_MASTER_C;
+				else if (strncmp(str, "i2c_bus_d", 9) == 0)
+					pdata->i2c_bus = AML_I2C_MASTER_D;
+				else if (strncmp(str, "i2c_bus_ao", 10) == 0)
+					pdata->i2c_bus = AML_I2C_MASTER_AO;
+				else
+					pdata->i2c_bus = AML_I2C_MASTER_A; 
+			}
+			DBG_PRINT("%s: i2c_bus=%s[%d]\n", pdata->name, str, pdata->i2c_bus);
+			break;
+		case LCD_EXTERN_SPI:
+			break;
+		default:
+			break;
+	}
+	
+	return 0;
+}
+
+static int lcd_extern_probe(struct platform_device *pdev)
+{
+	struct device_node* child;
+	struct i2c_board_info i2c_info;
+	struct i2c_adapter *adapter;
+	struct i2c_client *i2c_client;
+	int i = 0;
+	struct lcd_extern_data_t *pdata = NULL;
+	
+	if (!pdata)
+		pdata = kzalloc(sizeof(*pdata), GFP_KERNEL);
+	if (!pdata)	{
+		printk("failed to alloc data\n");
+		goto lcd_extern_probe_failed;
+	}
+	
+	pdev->dev.platform_data = pdata;
+	
+	for_each_child_of_node(pdev->dev.of_node, child) {
+		if (get_lcd_extern_dt_data(child, pdata) != 0) {
+			printk("failed to get dt data\n");
+			goto lcd_extern_probe_failed;
+		}
+		if (pdata->status == 1) {
+			switch (pdata->type) {
+				case LCD_EXTERN_I2C:
+					memset(&i2c_info, 0, sizeof(i2c_info));
+					
+					adapter = i2c_get_adapter(pdata->i2c_bus);
+					if (!adapter) {
+						printk("warning£ºfailed to get i2c adapter\n");
+						goto lcd_extern_probe_failed;
+					}
+					
+					strncpy(i2c_info.type, pdata->name, I2C_NAME_SIZE);
+					i2c_info.addr = pdata->addr;
+					i2c_info.platform_data = pdata;
+					i2c_info.flags=0;
+					if(i2c_info.addr>0x7f)
+						i2c_info.flags=0x10;
+					i2c_client = i2c_new_device(adapter, &i2c_info);
+					if (!i2c_client) {
+						printk("%s :failed to new i2c device\n", pdata->name);
+						goto lcd_extern_probe_failed;
+					}
+					else{
+						DBG_PRINT("%s: new i2c device succeed\n",((struct lcd_extern_data_t *)(i2c_client->dev.platform_data))->name);
+					}
+					
+					lcd_ext_driver.type = LCD_EXTERN_I2C;
+					break;
+				case LCD_EXTERN_SPI:
+					lcd_ext_driver.type = LCD_EXTERN_SPI;
+					break;
+				default:
+					break;
+			}
+			goto lcd_extern_probe_successful;	//only 1 driver allowed
+		}
+		i++;
+	}
+
+lcd_extern_probe_successful:
+	printk("%s ok\n", __FUNCTION__);
+	return 0;
+	
+lcd_extern_probe_failed:
+	if (pdata)
+		kfree(pdata);
+	return -1;
+}
+
+static int lcd_extern_remove(struct platform_device *pdev)
+{
+	if (pdev->dev.platform_data)
+	 	kfree (pdev->dev.platform_data);
+    return 0;
+}
+
+#ifdef CONFIG_USE_OF
+static const struct of_device_id aml_lcd_extern_dt_match[]={
+	{	
+		.compatible = "amlogic,lcd_extern",
+	},
+	{},
+};
+#else
+#define aml_lcd_extern_dt_match NULL
+#endif
+
+static struct platform_driver aml_lcd_extern_driver = {
+	.probe		= lcd_extern_probe,
+	.remove		= lcd_extern_remove,
+	.driver		= {
+		.name	= "aml_lcd_extern",
+		.owner	= THIS_MODULE,
+#ifdef CONFIG_USE_OF
+		.of_match_table = aml_lcd_extern_dt_match,
+#endif
+	},
+};
+
+static int __init lcd_extern_init(void)
+{
+	int ret;
+	printk("%s\n", __FUNCTION__);
+	ret = platform_driver_register(&aml_lcd_extern_driver);
+	if (ret) {
+		printk("failed to register lcd extern driver module\n");
+		return -ENODEV;
+	}
+	return ret;
+}
+
+static void __exit lcd_extern_exit(void)
+{
+	platform_driver_unregister(&aml_lcd_extern_driver);
+}
+
+module_init(lcd_extern_init);
+module_exit(lcd_extern_exit);
+
+MODULE_LICENSE("GPL v2");
+MODULE_DESCRIPTION("Amlogic LCD External bridge driver");
\ No newline at end of file
diff --git a/drivers/amlogic/display/vout/lcd_extern/tc101.c b/drivers/amlogic/display/vout/lcd_extern/tc101.c
new file mode 100755
index 000000000000..9ef5c608c8c9
--- /dev/null
+++ b/drivers/amlogic/display/vout/lcd_extern/tc101.c
@@ -0,0 +1,211 @@
+/*
+ * AMLOGIC lcd external driver.
+ *
+ * Communication protocol:
+ * I2C 
+ *
+ */
+
+#include <linux/kernel.h>
+#include <linux/init.h>
+#include <linux/module.h>
+#include <linux/slab.h>
+#include <linux/jiffies.h> 
+#include <linux/i2c.h>
+#include <linux/i2c-aml.h>
+#include <linux/miscdevice.h>
+#include <linux/mutex.h>
+#include <linux/mm.h>
+#include <linux/device.h>
+#include <linux/fs.h>
+#include <linux/delay.h>
+#include <linux/sysctl.h>
+#include <asm/uaccess.h>
+#include <mach/pinmux.h>
+#include <mach/gpio.h>
+#include <linux/platform_device.h>
+#include <linux/amlogic/vout/aml_lcd_extern.h>
+
+static struct i2c_client *aml_tc101_client;
+extern struct aml_lcd_extern_driver_t lcd_ext_driver;
+
+#define LCD_EXTERN_NAME			"lcd_tc101"
+
+static unsigned char tc101_init_table[][3] = {
+	//{0xff, 0xff, 20},//delay mark(20ms)
+	{0xf8, 0x30, 0xb2},
+	{0xf8, 0x33, 0xc2},
+	{0xf8, 0x31, 0xf0},
+	{0xf8, 0x40, 0x80},
+	{0xf8, 0x81, 0xec},
+	{0xff, 0xff, 0xff},//end mark
+};
+
+static int aml_lcd_i2c_write(struct i2c_client *i2client,unsigned char *buff, unsigned len)
+{
+    int res = 0;
+    struct i2c_msg msg[] = {
+        {
+        .addr = i2client->addr,
+        .flags = 0,
+        .len = len,
+        .buf = buff,
+        }
+    };
+	
+    res = i2c_transfer(i2client->adapter, msg, 1);
+    if (res < 0) {
+        printk("%s: i2c transfer failed [addr 0x%02x]\n", __FUNCTION__, i2client->addr);
+    }
+    
+    return res;
+}
+
+static int aml_lcd_i2c_read(struct i2c_client *i2client,unsigned char *buff, unsigned len)
+{
+    int res = 0;
+    struct i2c_msg msgs[] = {
+        {
+            .addr = i2client->addr,
+            .flags = 0,
+            .len = 1,
+            .buf = buff,
+        },
+        {
+            .addr = i2client->addr,
+            .flags = I2C_M_RD,
+            .len = len,
+            .buf = buff,
+        }
+    };
+    res = i2c_transfer(i2client->adapter, msgs, 2);
+    if (res < 0) {
+        printk("%s: i2c transfer failed [addr 0x%02x]\n", __FUNCTION__, i2client->addr);
+    }
+
+    return res;
+}
+
+static int tc101_reg_read(unsigned char reg, unsigned char *buf)
+{
+	int ret=0;
+
+	return ret;
+}
+
+static int tc101_reg_write(unsigned char reg, unsigned char value)
+{
+	int ret=0;
+	
+	return ret;
+}
+
+static int tc101_init(void)
+{
+	unsigned char tData[4];
+	int i=0, end_mark=0;
+	int ret=0;
+	
+	while (end_mark == 0) {
+		if ((tc101_init_table[i][0] == 0xff) && (tc101_init_table[i][1] == 0xff)) {	//special mark
+			if (tc101_init_table[i][2] == 0xff) {	//end mark
+				end_mark = 1;
+			}
+			else {	//delay mark
+				mdelay(tc101_init_table[i][2]);
+			}
+		}
+		else {
+			tData[0]=tc101_init_table[i][0];
+			tData[1]=tc101_init_table[i][1];
+			tData[2]=tc101_init_table[i][2];
+			aml_lcd_i2c_write(aml_tc101_client, tData, 3);
+		}
+		i++;
+	}
+	printk("%s\n", __FUNCTION__);
+	return ret;
+}
+
+static int tc101_remove(void)
+{
+	int ret=0;
+	
+	return ret;
+}
+
+static int aml_lcd_extern_driver_update(struct aml_lcd_extern_driver_t* lcd_ext)
+{
+	lcd_ext->name = LCD_EXTERN_NAME;
+	lcd_ext->reg_read = tc101_reg_read;
+	lcd_ext->reg_write = tc101_reg_write;
+	lcd_ext->power_on = tc101_init;
+	lcd_ext->power_off = tc101_remove;
+	
+	return 0;
+}
+
+static int aml_tc101_probe(struct i2c_client *client, const struct i2c_device_id *id)
+{
+	int ret;
+	
+	if (!i2c_check_functionality(client->adapter, I2C_FUNC_I2C)) {
+		printk("%s: functionality check failed\n", __FUNCTION__);
+		return -ENODEV;
+	} else {
+		aml_tc101_client = client;
+		aml_lcd_extern_driver_update(&lcd_ext_driver);
+	}
+
+	printk("%s OK\n", __FUNCTION__);
+
+	return 0;
+}
+
+static int aml_tc101_remove(struct i2c_client *client)
+{
+	return 0;
+}
+
+static const struct i2c_device_id aml_tc101_id[] = {
+	{LCD_EXTERN_NAME, 0},
+	{ }
+};
+MODULE_DEVICE_TABLE(i2c, aml_tc101_id);
+
+static struct i2c_driver aml_tc101_driver = {
+	.probe    = aml_tc101_probe,
+	.remove   = aml_tc101_remove,
+	.id_table = aml_tc101_id,
+	.driver = {
+		.name = LCD_EXTERN_NAME,
+		.owner =THIS_MODULE,
+	},
+};
+
+static int __init aml_tc101_init(void)
+{
+	int ret = 0;
+	//printk("%s\n", __FUNCTION__);
+
+	ret = i2c_add_driver(&aml_tc101_driver);
+	if (ret) {
+		printk("failed to register aml_tc101_driver\n");
+		return -ENODEV;
+	}
+
+	return ret;
+}
+
+static void __exit aml_tc101_exit(void)
+{
+	i2c_del_driver(&aml_tc101_driver);
+}
+
+//late_initcall(aml_tc101_init);
+module_init(aml_tc101_init);
+module_exit(aml_tc101_exit);
+
+MODULE_AUTHOR("AMLOGIC");
+MODULE_DESCRIPTION("LCD Extern driver for TC101");
+MODULE_LICENSE("GPL");
diff --git a/drivers/amlogic/display/vout/lcdoutc.c b/drivers/amlogic/display/vout/lcdoutc.c
index 8354d6ea7f60..306f201b031e 100755
--- a/drivers/amlogic/display/vout/lcdoutc.c
+++ b/drivers/amlogic/display/vout/lcdoutc.c
@@ -55,6 +55,9 @@
 #include "../backlight/aml_lcd_bl.h"
 #include "edp_drv.h"
 #include "mipi_dsi_util.h"
+#ifdef CONFIG_AML_LCD_EXTERN
+#include <linux/amlogic/vout/aml_lcd_extern.h>
+#endif
 #ifdef CONFIG_AMLOGIC_BOARD_HAS_PMU
 #include <linux/amlogic/aml_pmu_common.h>
 #endif
@@ -71,7 +74,7 @@ extern unsigned int clk_util_clk_msr(unsigned int clk_mux);
 #endif
 
 #define PANEL_NAME		"panel"
-#define DRIVER_DATE		"20131203"
+#define DRIVER_DATE		"20131230"
 #define DRIVER_VER		"310"
 
 //#define LCD_DEBUG_INFO
@@ -394,6 +397,9 @@ static void lcd_power_ctrl(Bool_t status)
 #ifdef CONFIG_AMLOGIC_BOARD_HAS_PMU
 	struct aml_pmu_driver *pmu_driver;
 #endif
+#ifdef CONFIG_AML_LCD_EXTERN
+	struct aml_lcd_extern_driver_t *lcd_extern_driver;
+#endif
 
 	DBG_PRINT("%s(): %s\n", __FUNCTION__, (status ? "ON" : "OFF"));
 	if (status) {
@@ -431,7 +437,17 @@ static void lcd_power_ctrl(Bool_t status)
 					lcd_ports_ctrl(ON);
 					break;
 				case LCD_POWER_TYPE_INITIAL:
-					printk("lcd power ctrl ON step %d lcd_init function is to be done.\n", i+1);
+#ifdef CONFIG_AML_LCD_EXTERN
+					lcd_extern_driver = aml_lcd_extern_get_driver();
+					if (lcd_extern_driver == NULL) {
+						printk("no lcd_extern driver\n");
+					}
+					else {
+						if (lcd_extern_driver->power_on)
+							lcd_extern_driver->power_on();
+						DBG_PRINT("%s power on\n", lcd_extern_driver->name);
+					}
+#endif
 					break;
 				default:
 					printk("lcd power ctrl ON step %d is null.\n", i+1);
@@ -477,7 +493,17 @@ static void lcd_power_ctrl(Bool_t status)
 					lcd_ports_ctrl(OFF);
 					break;
 				case LCD_POWER_TYPE_INITIAL:
-					printk("lcd power ctrl OFF step %d lcd_init function is to do.\n", i+1);
+#ifdef CONFIG_AML_LCD_EXTERN
+					lcd_extern_driver = aml_lcd_extern_get_driver();
+					if (lcd_extern_driver == NULL) {
+						printk("no lcd_extern driver\n");
+					}
+					else {
+						if (lcd_extern_driver->power_off)
+							lcd_extern_driver->power_off();
+						DBG_PRINT("%s power on\n", lcd_extern_driver->name);
+					}
+#endif
 					break;
 				default:
 					printk("lcd power ctrl OFF step %d is null.\n", i+1);
@@ -4628,7 +4654,6 @@ static int lcd_probe(struct platform_device *pdev)
 #endif
 
     pDev = (lcd_dev_t *)kmalloc(sizeof(lcd_dev_t), GFP_KERNEL);
-
     if (!pDev) {
         printk("[lcd probe]: Not enough memory.\n");
         return -ENOMEM;
diff --git a/include/linux/amlogic/vout/aml_lcd_extern.h b/include/linux/amlogic/vout/aml_lcd_extern.h
new file mode 100755
index 000000000000..d86569d5d776
--- /dev/null
+++ b/include/linux/amlogic/vout/aml_lcd_extern.h
@@ -0,0 +1,35 @@
+
+#ifndef __AMLOGIC_LCD_EXTERN_H_
+#define __AMLOGIC_LCD_EXTERN_H_
+
+typedef enum {
+	LCD_EXTERN_I2C = 0,
+	LCD_EXTERN_SPI,
+	LCD_EXTERN_MAX,
+} Lcd_Extern_Type_t;
+
+struct aml_lcd_extern_driver_t {
+	char *name;
+	Lcd_Extern_Type_t type;
+	int  (*reg_read)  (unsigned char reg, unsigned char *buf);
+    int  (*reg_write) (unsigned char reg, unsigned char value);
+	int  (*power_on)(void);
+	int  (*power_off)(void);
+	// void  (*fun1_no_para)(void);
+	// void  (*fun2_no_para)(void);
+	// void  (*fun1_para)(unsigned para);
+	// void  (*fun2_para)(unsigned para); 
+};
+
+struct lcd_extern_data_t {
+	char *name;
+	Lcd_Extern_Type_t type;
+	int status;
+	int  i2c_bus;
+	int  addr;
+};
+
+extern struct aml_lcd_extern_driver_t* aml_lcd_extern_get_driver(void);
+
+#endif
+
-- 
2.19.0

